src/lib.rs

```diff
@@ -52,6 +51,58 @@ fn repeat_byte(b: u8) -> usize {
     rep
 }
 
+/// An iterator for memchr
+pub struct Memchr<'a> {
+    needle: u8,
+    // The haystack to iterate over
+    haystack: &'a [u8],
+    // The index
+    position: usize,
+}
+
+impl<'a> Memchr<'a> {
+    /// Create a new Memchr that's initalized to zero with a haystack
```

Comment:
I think the docs could be a little better. In particular, "initialized to zero" seems like an implementation detail that doesn't belong in the public API. What about this?

> Creates a new iterator that yields all positions of `needle` in `haystack`.

Similarly for the other iterators.